### PR TITLE
Fix freeze on non-thread-safe custom importers

### LINKED ISSUE #85237: Freeze on ResourceSaver.save in custom importer when using multiple threads
### Godot version

4.2.rc1

### System information

Godot v4.2.rc1 - Manjaro Linux #1 SMP PREEMPT_DYNAMIC Thu Nov  9 04:27:49 UTC 2023 - X11 - Vulkan (Forward+) - integrated AMD Radeon Vega 8 Graphics (RADV RAVEN) () - AMD Ryzen 5 3550H with Radeon Vega Mobile Gfx (8 Threads)

### Issue description

Using a custom importer on a lot of files, with import threads enabled, sometimes causes Godot to fully freeze (including the UI), with no error in stdout/stderr.

Importing each file individually never freezes (and they all import correctly). Disabling importer threads seems to fully solve the problem.

Even setting a minimal project with only one custom importer and 10 files, it happens to freeze during import.

In this minimal project, it does not always happen, but randomly under certain conditions:

- `.godot/imported` is empty
- `*.TXTR.import` files do not exist yet
- The importer plugin is enabled in Project Settings for the first time (interestingly it happens less if it's already enabled in `project.godot`)

The UI freezes fully like this:

image

Here is the minimal project:

[bug.tar.gz](https://github.com/godotengine/godot/files/13374254/bug.tar.gz)

The output before freeze (from print statements in the plugin):

```
Godot Engine v4.2.beta6.official.64150060f - https://godotengine.org
Vulkan API 1.3.257 - Forward+ - Using Vulkan Device #0: AMD - AMD Radeon Vega 8 Graphics
 
WARNING: Blend file import is enabled in the project settings, but no Blender path is configured in the editor settings. Blend files will not be imported.
     at: _editor_init (modules/gltf/register_types.cpp:63)
res://data/M000.TXTR -> res://.godot/imported/M000.TXTR-4bb0fc982e0e8da4bb882b98bceb0aad.res
res://data/M001.TXTR -> res://.godot/imported/M001.TXTR-fe74a5ef0315d8fdb6f9bdc09f3f6428.res
res://data/M003.TXTR -> res://.godot/imported/M003.TXTR-967cf8977babfd44b3b1c3d40e57708c.res
res://data/M005.TXTR -> res://.godot/imported/M005.TXTR-06dad9b17abe199d2512f9c3387587af.res
res://data/M002.TXTR -> res://.godot/imported/M002.TXTR-83aa75747fe3ba7170d1863f4b39e98f.res
res://data/M004.TXTR -> res://.godot/imported/M004.TXTR-b19fd80f2afb1ffc7b8f131016cd8ab9.res
res://data/M006.TXTR -> res://.godot/imported/M006.TXTR-d45d4916580d73bdbed1c55d81a92f01.res
res://data/M007.TXTR -> res://.godot/imported/M007.TXTR-81e6b1cfd963935c1849355e686ea70f.res
res://data/M009.TXTR -> res://.godot/imported/M009.TXTR-f290dc67f49ba13700c462d305645eb5.res
res://data/M008.TXTR -> res://.godot/imported/M008.TXTR-6b69818b71410aede94f9d288f0be74f.res
```

The state of import files after freezing:

```
~/w/g/bug > find . -name '*.import' | sort                                                                                                                                                                        
./data/M000.TXTR.import
./data/M001.TXTR.import
./data/M002.TXTR.import
./data/M003.TXTR.import
./data/M004.TXTR.import
./data/M005.TXTR.import
./data/M006.TXTR.import
./data/M007.TXTR.import
./data/M009.TXTR.import
./icon.svg.import
```

The state of `.godot/imported` after freezing:

```
~/w/g/bug > ll .godot/imported                                                                                                                                                                      

   3288 16 nov.  09:20 icon.svg-218a8f2b3041327d8a5756f3a245f83b.ctex
     91 16 nov.  09:20 icon.svg-218a8f2b3041327d8a5756f3a245f83b.md5
     91 16 nov.  09:20 M000.TXTR-4bb0fc982e0e8da4bb882b98bceb0aad.md5
  49617 16 nov.  09:20 M000.TXTR-4bb0fc982e0e8da4bb882b98bceb0aad.res
     91 16 nov.  09:20 M001.TXTR-fe74a5ef0315d8fdb6f9bdc09f3f6428.md5
  49617 16 nov.  09:20 M001.TXTR-fe74a5ef0315d8fdb6f9bdc09f3f6428.res
     91 16 nov.  09:20 M002.TXTR-83aa75747fe3ba7170d1863f4b39e98f.md5
  49617 16 nov.  09:20 M002.TXTR-83aa75747fe3ba7170d1863f4b39e98f.res
     91 16 nov.  09:20 M003.TXTR-967cf8977babfd44b3b1c3d40e57708c.md5
  49617 16 nov.  09:20 M003.TXTR-967cf8977babfd44b3b1c3d40e57708c.res
     91 16 nov.  09:20 M004.TXTR-b19fd80f2afb1ffc7b8f131016cd8ab9.md5
  12753 16 nov.  09:20 M004.TXTR-b19fd80f2afb1ffc7b8f131016cd8ab9.res
     91 16 nov.  09:20 M005.TXTR-06dad9b17abe199d2512f9c3387587af.md5
 262610 16 nov.  09:20 M005.TXTR-06dad9b17abe199d2512f9c3387587af.res
     91 16 nov.  09:20 M006.TXTR-d45d4916580d73bdbed1c55d81a92f01.md5
 197073 16 nov.  09:20 M006.TXTR-d45d4916580d73bdbed1c55d81a92f01.res
     91 16 nov.  09:20 M007.TXTR-81e6b1cfd963935c1849355e686ea70f.md5
 197073 16 nov.  09:20 M007.TXTR-81e6b1cfd963935c1849355e686ea70f.res
      0 16 nov.  09:20 M008.TXTR-6b69818b71410aede94f9d288f0be74f.res-oHioUv
     91 16 nov.  09:20 M009.TXTR-f290dc67f49ba13700c462d305645eb5.md5
  12753 16 nov.  09:20 M009.TXTR-f290dc67f49ba13700c462d305645eb5.res
```

_Originally posted by @thunderk in https://github.com/godotengine/godot/issues/34961#issuecomment-1814002433_

### Steps to reproduce

* Open minimal project above
* Enable addon in Project Settings
* If importing succeeds, close project, delete project, do it all again, until it freezes (it's very random)

### Minimal reproduction project

[bug.tar.gz](https://github.com/godotengine/godot/files/13374254/bug.tar.gz)

### PR DESCRIPTION
**Pull request by hunterloftis**
Godot currently assumes that all importers are thread-safe, leading to non-deterministic, OS and hardware-dependent deadlocks. Several of Godot's built-in importers already override this false assumption in order to avoid freezing the editor:

- https://github.com/godotengine/godot/blob/92e51fca7247c932f95a1662aefc28aca96e8de6/editor/import/3d/resource_importer_obj.h#L66-L67
- https://github.com/godotengine/godot/blob/92e51fca7247c932f95a1662aefc28aca96e8de6/editor/import/3d/resource_importer_scene.h#L307

This updates the `EditorImportPlugin` implementation and documentation to make custom importers safe to use by default, while still allowing them to opt into threaded concurrency optimizations _if_ they are actually thread-safe.

Fixes https://github.com/godotengine/godot/issues/85237

### CODE REVIEW

**Comment by hunterloftis** [comment]
If this general approach is :+1: from the core team, I'll go through and enable threaded imports in the built-in importers that are currently thread-safe & using the `true` default.

**Comment by hunterloftis** [comment]
> I think changing the meaning of methods in a stable release might not make sense. I have no idea when godot 5 will be released.

Could you clarify this? This PR doesn't change the meaning of any methods.

> Also importers can override this method so its a uiux bug not a blocker

Assuming they know about this bug. The reason multiple issues exist about this is because many users don't know that godot chooses the most dangerous possible default and may deadlock if they don't override this method:

- https://github.com/godotengine/godot/issues/85237
- https://github.com/godotengine/godot/issues/48265
- https://github.com/godotengine/godot/issues/85773
- https://github.com/godotengine/godot/issues/96242
- https://github.com/godotengine/godot/issues/48265
- https://github.com/godotengine/godot/issues/94586

Further, since the deadlock is host-hardware and OS dependent and random, even testing by multiple users may not reveal an underlying issue. I discovered it for example after integrating [godot-ink](https://github.com/paulloz/godot-ink) _when our first Windows-based developer tried to run the project_.

> Edited: doing the backwards compatibility of deprecrating the old method and renaming may work but not sure.

Again, could you clarify? There is no "old method" or method renaming here.

Currently, all custom importers for Godot fall into one of three categories:

1. Works, with concurrency optimizations
2. Deadlocks
3. Author overrides [_can_import_threaded](https://docs.godotengine.org/en/stable/classes/class_editorimportplugin.html#class-editorimportplugin-private-method-can-import-threaded) so it works, without concurrency optimizations

This change removes the deadlock category, so:

1. Works, without concurrency optimizations
2. Author overrides [_can_import_threaded](https://docs.godotengine.org/en/stable/classes/class_editorimportplugin.html#class-editorimportplugin-private-method-can-import-threaded) so it continues to work, with concurrency optimizations.

**Comment by fire** [comment]
@dsnopek Is there any problems with changing this definition in gdextension? I can see the merits but curious

**Comment by hunterloftis** [comment]
@AThousandShips I don't know the metrics used for "breaks compat" in this project, but in terms of functionality this change is fully backwards-compatible for both internal and custom importers.

**Comment by AThousandShips** [comment]
Changing the default return of a virtual method breaks compatibility in my book, it makes it so that anyone who didn't override this method because they were just going to return `true` now has broken code, how is it backwards compatible at all? Granted the loss of threaded loading might not be critical but still

But let's see what dsnopek says

**Comment by hunterloftis** [comment]
Like I said, the project can of course use whatever metric it likes for "breaks compat." I'm clarifying for accuracy, since this is not true:

> it makes it so that anyone who didn't override this method because they were just going to return true now has **broken** code

Their code will not be broken, any more than their code would be broken running on a single-core CPU (as the behavior of their code will be identical to how it would run today on such a CPU - which, in all cases, is _more_ likely to be stable).

**Comment by clayjohn** [comment]
I don't think this breaks compatibility. Yes it changes default behaviour, but that doesn't mean compatibility is broken. 

IMO "breaks compat" should be reserved for cases where we actually change or remove a feature in a way that breaks things. 

In this case, the change in behaviour won't break anything, at most it will make your custom resource import slower.

**Comment by AThousandShips** [comment]
I'd say that the public API is generally considered part of compatibility, especially behavior changes, but I'll defer to dsnopek on this

Either way I'd say this should be documented in the upgrading from 4.3 to 4.4 page when that is made

**Review by fire — APPROVED** [review]
I approve this because we switch from an optimistic default assumption that custom importers are thread-safe to a conservative default assumption that they are not thread-safe. 

@dsnopek mentions this is an conservative default that changes the meaning but is safe in the sense it doesn't destroy code functionality, but only run slower.

Custom importers that are thread-safe can be declared thread-safe by their code maintainers.

> This updates the EditorImportPlugin implementation and documentation to make custom importers safe to use by default while allowing them to opt into threaded concurrency optimizations if they are thread-safe.

We could ensure the status quo is kept for existing importers in our tree.

@lyuma @TokageItLab @SaracenOne might have opinions, but I'm ok with making this quality-of-life change where it's easier to implement custom importers without crashing because of a lack of knowledge of the threaded requirement - I assume this was because previous godot engine versions didn't have gdextension class extension and was an optimistic way to encourage godot engine contributors to test threaded-ness in their importer code by default when it was implemented by @reduz for Godot Engine 4.0.

**Comment by hunterloftis** [comment]
Is there anything I need to do to have this ready for 4.4 or is it good-to-go, just waiting?

**Review by adamscott — APPROVED** [review]
Seems ok to me. I was worried it could impact single-threaded builds, but it doesn't seem so, especially since the default for `can_import_threaded()` was `true`.

**Comment by Repiteo** [comment]
Thanks! Congratulations on your first contribution! 🎉

**Pull request merged**
